prusti-common/src/vir/optimizations/predicates/delete_unused_predicates.rs

```diff
@@ -0,0 +1,191 @@
+    let mut collector = UsedPredicateCollector::new();
+    super::walk_methods(methods, &mut collector);
+    super::walk_functions(functions, &mut collector);
+
+    // DeadBorrowToken$ is a used predicate but it does not appear in VIR becaue it is only created when viper code is created from VIR
+    collector
+        .used_predicates
+        .insert("DeadBorrowToken$".to_string());
+    collector.used_predicates
+}
+
+fn get_used_predicates_in_predicates(predicates: &[Predicate]) -> BTreeSet<String> {
+    let mut collector = UsedPredicateCollector::new();
+
+    for pred in predicates {
+        match pred {
+            Predicate::Struct(StructPredicate { body: Some(e), .. }) => {
+                ExprWalker::walk(&mut collector, e)
+            }
+            Predicate::Struct(_) => { /* ignore */ }
+            Predicate::Enum(p) => {
+                ExprWalker::walk(&mut collector, &p.discriminant);
+                ExprWalker::walk(&mut collector, &p.discriminant_bounds);
+
+                for (e, _, sp) in &p.variants {
+                    ExprWalker::walk(&mut collector, e);
+                    collector.used_predicates.insert(sp.name.to_string());
+                    sp.body
+                        .iter()
+                        .for_each(|e| ExprWalker::walk(&mut collector, e))
+                }
+            }
+            Predicate::Bodyless(_, _) => { /* ignore */ }
+        }
+    }
+    collector.used_predicates
+}
+
+fn remove_body_of_predicates_if_possible(
+    predicates: &[Predicate],
+    predicates_only_used_in_predicates: &BTreeSet<String>,
+) -> Vec<Predicate> {
+    let mut new_predicates = predicates.to_vec();
+
+    new_predicates.iter_mut().for_each(|predicate| {
+        let predicates_used_in_this_predicate =
+            get_used_predicates_in_predicates(&[predicate.clone()]);
+        if predicates_used_in_this_predicate
+            .intersection(&predicates_only_used_in_predicates)
+            .next()
+            .is_some()
+        {
+            if let Predicate::Struct(sp) = predicate {
+                sp.body = None;
+            }
+        }
+    });
+    new_predicates
+}
+
+pub fn delete_unused_predicates(
+    methods: &[CfgMethod],
+    functions: &[Function],
+    predicates: &[Predicate],
+) -> Vec<Predicate> {
+    let mut has_changed = true;
+    let mut new_predicates: Vec<Predicate> = predicates.to_vec();
+
+    let used_preds = get_used_predicates(methods, functions);
+
+    debug!(
+        "The used predicates in functions and methods are {:?}",
+        &used_preds
+    );
+
+    while has_changed {
+        has_changed = false;
+
+        let predicates_used_in_predicates = get_used_predicates_in_predicates(&new_predicates);
+        debug!(
+            "The used predicates in predicates are {:?}",
+            &predicates_used_in_predicates
+        );
+        new_predicates = new_predicates
+            .into_iter()
+            .filter(|p| {
```

Comment:
You probably could use [`Vec::retain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain) here.